### PR TITLE
feat(common): ldml: error messages from the compiler   🙀

### PR DESCRIPTION
**Pull request by srl295**
- more refactor into common/web/types
- refactor test harness
- move throw into Errors

For: #7534 / #7964

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts [hunk 1]
```diff
@@ -154,14 +155,16 @@ export default class LDMLKeyboardXMLSourceFileReader {
...
+        for (let err of ajv.errors) {
+          // report each err
+          this.callbacks.reportMessage(CommonTypesMessages.Fatal_AJVError(err))
+        }
       } else {
         // Not a single error, so fall through to errorsText()
-        throw new Error(ajv.errorsText());
+        this.callbacks.reportMessage(CommonTypesMessages.Fatal_UnknownXMLValidationError({
+          errorsText: ajv.errorsText() || 'Unknown AJV Error'
+        }));
```
Comment (mcdurdin) [inline common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts:167]:
```suggestion
        this.callbacks.reportMessage(CommonTypesMessages.Error_UnknownXMLValidationError({
          errorsText: ajv.errorsText() || 'Unknown AJV Error'
        }));
```

File: common/web/types/src/util/common-events.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+const SevFatal = CompilerErrorSeverity.Fatal | KmcErrMask;
+
+export class CommonTypesMessages {
+  // structured Ajv validation error
+  static Fatal_AJVError = (o:ErrorObject) => m(this.FATAL_XMLValidationError,
+    `${o.instancePath}: ${o.keyword}: ${o.message} ${JSON.stringify(o.params||{})}`);
+  // some other unknown Ajv error. Same as above, but without structure
+  static Fatal_UnknownXMLValidationError = (o:{errorsText: string}) => m(this.FATAL_XMLValidationError,
+    `Validation failed: ${o.errorsText}`);
+  static FATAL_XMLValidationError = SevFatal | 0x0001;
```
Comment (mcdurdin) [inline common/web/types/src/util/common-events.ts:20]: A few things:

1. `${JSON.stringify(o.params||{})}` needs to be in the caller, to keep the message constants translateable (complex embedded expressions are going to cause grief for translators).
2. These should be ERRORs, not FATALs -- FATALs are reserved for compiler bugs and out of memory.
4. Each error string should be paired with a unique error code -- even though these two errors come from the same code area, we are making this a pattern for ease of maintenance and consistency.
5. The error message needs to be interpretable by a keyboard developer. I'm unclear what is going to be returned by AJV for some of these fields?

```suggestion
  static Error_AJVError = (o:{instancePath:string, keyword:string, message: string, params: string}) => m(this.ERROR_AJVError,
    `Error validating LDML XML file: ${o.instancePath}: ${o.keyword}: ${o.message} ${o.params}`);
  static ERROR_AJVError = SevError | 0x0001;

  // some other unknown Ajv error. Same as above, but without structure
  static Error_UnknownXMLValidationError = (o:{errorsText: string}) => m(this.ERROR_UnknownXMLValidationError,
    `Unknown error validating LDML XML file: ${o.errorsText}`);
  static ERROR_UnknownXMLValidationError = SevError | 0x0002;
```

File: common/web/types/src/util/common-events.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import { ErrorObject } from "ajv";
+
+const m = (code: number, message: string) : CompilerEvent => { return { code, message } };
+
+const KmcErrMask   = CompilerErrorNamespace.Types;
+// const SevInfo = CompilerErrorSeverity.Info | KmcErrMask;
+// const SevHint = CompilerErrorSeverity.Hint | KmcErrMask;
+// const SevWarn = CompilerErrorSeverity.Warn | KmcErrMask;
+// const SevError = CompilerErrorSeverity.Error | KmcErrMask;
+const SevFatal = CompilerErrorSeverity.Fatal | KmcErrMask;
```
Comment (mcdurdin) [inline common/web/types/src/util/common-events.ts:11]:
```suggestion
const TypesErrMask   = CompilerErrorNamespace.Types;
// const SevInfo = CompilerErrorSeverity.Info | TypesErrMask;
// const SevHint = CompilerErrorSeverity.Hint | TypesErrMask;
// const SevWarn = CompilerErrorSeverity.Warn | TypesErrMask;
const SevError = CompilerErrorSeverity.Error | TypesErrMask;
// const SevFatal = CompilerErrorSeverity.Fatal | TypesErrMask;
```

File: common/web/types/src/util/compiler-interfaces.ts [hunk 4]
```diff
@@ -6,6 +6,27 @@ export interface CompilerEvent {
...
+  Fatal =         0x400000,
+  Severity_Mask = 0xF00000, // includes reserved bits
+  Error_Mask =    0x0FFFFF,
+};
+
+export enum CompilerErrorNamespace {
+  /**
+   * KMC errors between 0x0000…0x0FFF
+   */
+  Kmc = 0x0000,
```
Comment (mcdurdin) [inline common/web/types/src/util/compiler-interfaces.ts:23]: This should probably be `KeyboardCompiler` or perhaps (I'd prefer not) `KmcKeyboard`, not `Kmc`.

```suggestion
  /**
   * kmc-keyboard errors between 0x0000…0x0FFF
   */
  KeyboardCompiler = 0x0000,
```

File: common/web/types/src/util/compiler-interfaces.ts [hunk 5]
```diff
@@ -6,6 +6,27 @@ export interface CompilerEvent {
...
+  /**
+   * common/web/types errors between 0x1000…0x1FFF
+   */
+  Types = 0x1000,
```
Comment (mcdurdin) [inline common/web/types/src/util/compiler-interfaces.ts:27]: Maybe `CommonTypes` instead of just `Types`?

File: developer/src/kmc-keyboard/src/compiler/messages.ts [hunk 6]
```diff
@@ -1,20 +1,14 @@
...
-const m = (code: number, message: string) => { return { code, message } };
-const SevInfo = CompilerErrorSeverity.Info;
-const SevHint = CompilerErrorSeverity.Hint;
-// const SevWarn = CompilerErrorSeverity.Warn;
-const SevError = CompilerErrorSeverity.Error;
-const SevFatal = CompilerErrorSeverity.Fatal;
+
+const m = (code: number, message: string) : CompilerEvent => { return { code, message } };
+
+const KmcErrMask   = CompilerErrorNamespace.Kmc;
```
Comment (mcdurdin) [inline developer/src/kmc-keyboard/src/compiler/messages.ts:6]:
```suggestion
const KmcErrMask   = CompilerErrorNamespace.KeyboardCompiler;
```

File: common/web/types/src/util/common-events.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { CompilerErrorNamespace, CompilerErrorSeverity, CompilerEvent } from './compiler-interfaces.js';
+import { ErrorObject } from "ajv";
+
+const m = (code: number, message: string) : CompilerEvent => { return { code, message } };
```
Comment (mcdurdin) [inline common/web/types/src/util/common-events.ts:4]: This could be imported from `compiler-interfaces.js` thusly, and thereby share the definition:

```js
import { CompilerMessageSpec as m } from './compiler-interfaces.js';
```

**Review by mcdurdin — APPROVED** [review]
LGTM. Good to see some of these interfaces solidifying. Some tweaks suggested.

**Comment by mcdurdin** [comment]
This is great -- we'll work from this base in improving the package compiler error reporting.

File: common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts [hunk 8]
```diff
@@ -154,14 +155,16 @@ export default class LDMLKeyboardXMLSourceFileReader {
...
       // Try to improve the message
-      if (ajv.errors?.length === 1) {
-        // Only one error. Try to improve the message.
-        const err = ajv.errors[0];
-        const { instancePath, keyword, params, message } = err;
-        throw new Error(`${instancePath}: ${keyword}: ${message} ${JSON.stringify(params||{})}`);
+      if (ajv.errors?.length > 0) {
+        for (let err of ajv.errors) {
+          // report each err
+          this.callbacks.reportMessage(CommonTypesMessages.Fatal_AJVError(err))
```
[the remark this answers is not in the document]
Comment (srl295) [reply common/web/types/src/ldml-keyboard/ldml-keyboard-xml-reader.ts:161]: i'l add a comment on the categories

File: common/web/types/src/util/common-events.ts (hunk 2 above)
Comment (srl295) [reply common/web/types/src/util/common-events.ts:20]: 1. good point. 

2. +1

3. makes sense. but related to point 4

4. "what is going to be returned by AJV?"  

There is https://github.com/ajv-validator/ajv-errors which can go really deep with customizing ajv's errors, probably more than we need here.  

The value of `params` is specified [here](https://ajv.js.org/api.html#error-objects) - looks like everything is 1 level deep key=value, that's why I went with stringify.  This might read a little better:

```shell
> Object.entries({z: 123, a: 123, b: "foo", c: "bar baz"}).sort().map(([k,v])=>`${k}="${v}"`).join(' ')
'a="123" b="foo" c="bar baz" z="123"'
```

Basically the point here is:  validation failed.  Here's some context from AJV.  We could add some text such as `please double check your XML structure`?

File: common/web/types/src/util/compiler-interfaces.ts (hunk 5 above)
Comment (srl295) [reply common/web/types/src/util/compiler-interfaces.ts:27]: what do you have against Web, which is now excluded? 🕸️ :)

File: common/web/types/src/util/common-events.ts (hunk 7 above)
Comment (srl295) [reply common/web/types/src/util/common-events.ts:4]: +1

**Comment by srl295** [comment]
@mcdurdin going to re-request here to make sure the AJV directions sounds good

**Comment by srl295** [comment]
@mcdurdin  didn't get rebased properly. take a look now

**Comment by srl295** [comment]
@mcdurdin ready for re-review

File: common/web/types/src/util/common-events.ts (hunk 2 above)
Comment (mcdurdin) [reply common/web/types/src/util/common-events.ts]: Yep, no problem. On point 4 -- happy to return extra detail from AJV so long as there is a preamble "Unable to validate XML" or something like that.

File: common/web/types/src/util/common-events.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  static ERROR_SchemaValidationError = SevError | 0x0001;
+
+  static Error_ImportInvalidBase = (o: { base: string, path: string, subtag: string }) =>
+    m(this.ERROR_ImportInvalidBase,
+      `Import element with base ${o.base} is unsupported. Only ${constants.cldr_import_base} is supported.`);
+  static ERROR_ImportInvalidBase = SevError | 0x0002;
+
+  static Error_ImportInvalidPath = (o: { base: string, path: string, subtag: string }) =>
+    m(this.ERROR_ImportInvalidPath,
+      `import element with invalid path ${o.path}: expect the form '${constants.cldr_version_latest}./*.xml`);
```
Comment (mcdurdin) [inline common/web/types/src/util/common-events.ts:24]:
```suggestion
      `Import element with invalid path ${o.path}: expected the form '${constants.cldr_version_latest}./*.xml`);
```

File: common/web/types/src/util/common-events.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  static ERROR_ImportInvalidPath = SevError | 0x0003;
+
+  static Error_ImportReadFail = (o: { base: string, path: string, subtag: string }) =>
+    m(this.ERROR_ImportReadFail,
+      `Import could not read data with path ${o.path}: expect the form '${constants.cldr_version_latest}./*.xml'`);
```
Comment (mcdurdin) [inline common/web/types/src/util/common-events.ts:29]:
```suggestion
      `Import could not read data with path ${o.path}: expected the form '${constants.cldr_version_latest}./*.xml'`);
```

**Review by mcdurdin — APPROVED** [review]
LGTM
